### ISSUE TITLE
Not able to read /sys/kernel/security/tpm0/ascii_runtime_measurements 

### ISSUE THREAD

**Issue by ricardoas** [stefanberger/swtpm#61]
Hi, I just reproduced the install process @stenioaraujo described at https://github.com/stefanberger/swtpm/issues/4#issuecomment-385998296 (swtpm 2.0+QEMU+OVMF) but I ended up with no `pcrs `file inside `/sys/class/tpm/tpm0/device/` and no `tpm0` folder inside `/sys/kernel/security/` only an `ima` folder containing a single measurement for boot_aggregate.

After installing tpm2-tools from apt, I managed to print pcrs with `sudo resourcemgr & tpm2_listpcrs` and PCRs 0-7 are actually being used.

Have you ever experienced this behavior? I'd really appreciate if someone could point me in some direction to solve this... as I'm not even sure if it is a IMA problem or something I'm missing in wiring the emulator, qemu and OVMF.

**Comment by ricardoas** [comment]
I didn't know that not having a `pcrs` file inside sysfs was an expected behavior. 

The output of `dmesg | grep -i ima` shows:

```
$dmesg | grep -i ima
[    0.000000] Command line: BOOT_IMAGE=/boot/vmlinuz-4.4.0-122-generic root=LABEL=cloudimg-rootfs ro console=tty1 console=ttyS0
[    0.000000] kvm-clock: cpu 0, msr 0:3fe69001, primary cpu clock
[    0.000000] Kernel command line: BOOT_IMAGE=/boot/vmlinuz-4.4.0-122-generic root=LABEL=cloudimg-rootfs ro console=tty1 console=ttyS0
[    0.592002] evm: security.ima
[    3.492642] Allocating IMA MOK and blacklist keyrings.
[    4.456256] GPT:Primary header thinks Alt. header is not at the end of the disk.
[    5.283681] PM: Hibernation image not present or could not be loaded.
[   10.820360] systemd[1]: systemd 229 running in system mode. (+PAM +AUDIT +SELINUX +IMA +APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ -LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD -IDN)
```
So I'm assuming IMA is supported by the content of line with 10.820360. Right?

The output of dmesg | grep ACPI | grep TPM2 shows:
```
$dmesg | grep ACPI | grep TPM2
[    0.000000] ACPI: TPM2 0x000000003E728000 00004C (v04 BOCHS  BXPCTPM2 00000001 BXPC 00000001)
```
It seems to me I have a TPM 2 ACPI table... so I should have IMA measurements for boot path, right?

The problem I see now is that there is no `/sys/kernel/security/tpm0/ascii_runtime_measurements` file with the boot path measurements (PCRs 0 to 7) to recalculate and verify the `boot_aggregate` line at . `/sys/kernel/security/ima/ascii_runtime_measurements`. Do you know if I'm going in the wrong direction with IMA and TPM2? Does it log in another way when using TPM 2.0?

**Comment by stefanberger** [comment]
TPM 2 logs are written differently than TPM 1.2 logs. The SeaBIOS logs for TPM 2 wouldn't be displayed by the TPM subsystem of Linux. For EDK2 I think they are not displayed, either.

There's a blob here that is using a tool tpm2dumplog: https://blog.twobit.us/2016/05/31/tpm2-uefi-measurements-and-event-log/

The pointer to that tool doesn't work anymore. I am not sure what tool to use to display the TPM 2 log.

**Comment by stefanberger** [comment]
You need at least a 4.11 Linux kernel to get access to the tpm2 eventlog.

**Comment by ricardoas** [comment]
Thank you, @stefanberger ! I'll take a look on this post and see if I can find the tool...

I noticed that the guest VM was using ubuntu 16.04 with kernel older than 4.11. I'll give a try at 18.04 and see if it changes anything, but well... I guess I need to read a bit more to get a clearer idea on the TPM 2.0 way of doing things.

Thanks again and I'll update here if I get anything new...

**Comment by stefanberger** [comment]
@ricardoas Did updating to 18.04 help?

**Comment by ricardoas** [comment]
Hi @stefanberger, sorry for the delay. I updated to ubuntu 18.04 but I did not have the time to take a closer at tpm2dumplog tool, so I still have no access to logs. As I'm using StrongSwan to attest a server before authorizing it to enter a VPN and it works for TPM 1.2 using the binary measurements in `/sys/kernel/security/tpm0/`, I was hoping initially that it would work for TPM 2.0 the same way. But, StrongSwan is ignoring the lack of boot measurements for TPM 2.0, I hope I find time to investigate this soon.

Besides that, in the last weeks I noticed that TPM 2.0 PCRs 0-7 are zeroed when I use EDK2 with SecureBoot (what made me move back to SeaBIOS), but I could not identify yet if it is a problem on swtpm, EDK II, or even on the VM kernel modules. The only error I see in dmesg is:

```
[    2.290168] tpm_tis 00:06: 2.0 TPM (device-id 0x1, rev-id 1)
[    2.299051] tpm tpm0: A TPM error (256) occurred continue selftest
[    2.306930] tpm tpm0: starting up the TPM manually
```

Have you ever seen this kind of problem?

**Comment by stefanberger** [comment]
I installed Ubuntu 18.04 as well. I saw this error when running QEMU with an older version of SeaBIOS. Once I installed a recent version of SeaBIOS into /usr/share/qemu/bios-256k.bin, which includes the patch 'tpm: Request access to locality 0' with commit number `d1343e6863` this error disappeared when using the TPM 2 with a **CRB interface**. However, I don't get the directory `/sys/kernel/security/tpm0` but I have `/dev/tpm0` and `/dev/tpmrm0`, which is a bit odd. Both seem to work. 

Besides this IMA complains as follows:
```
[    1.485467] ima: No TPM chip found, activating TPM-bypass! (rc=-19)
```
-19 = -ENODEV. This is with kernel 4.15.0-29-generic.
Otherwise `dmesg` is clean of any other errors.

Using the 'tss2' package's tsspcrread shows that PCRs 0-7 have been extended by SeaBIOS:

```
> export TPM_INTERFACE_TYPE=dev
> tsspcrread -ha 0
count 1
 digest length 32
 d2 7c c1 26 14 b5 f4 ff 85 ed 10 94 95 e3 20 fb
 1e 54 95 eb 28 d5 07 e9 52 d5 10 91 e7 ae 2a 72
```

With a TPM 2 using the **TIS interface** IMA does NOT complain about no chip being found and extends PCR 10 as usual.

**Comment by ricardoas** [comment]
I dug a bit into edk2 and ovmf and I found out that PCRs 0-7 started being zeroed after tianocore/edk2@f15cb995bb3880b77e15afe6facd3da05e599a17. I sent an email on [their list](https://lists.01.org/pipermail/edk2-devel/2018-August/027838.html) to discover if this is a bug or an additional configuration needed when compiling ofmv. In the meantime, I'll set up a new environment with IMA and StrongSwan.

I also found out this mail https://lists.01.org/pipermail/edk2-devel/2018-June/025542.html starting a discussion on TPM2 ACPI table... I'm not sure but this might solve my problem with the measurements.
